include/gridtools/fn/sid_neighbor_table.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2022, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include <cstdint>
+#include <gridtools/common/array.hpp>
+#include <gridtools/fn/unstructured.hpp>
+#include <gridtools/sid/concept.hpp>
+#include <type_traits>
+
+namespace gridtools::fn::sid_neighbor_table {
+    namespace sid_neighbor_table_impl_ {
+        template <class IndexDimension, class NeighborDimension, int32_t MaxNumNeighbors, class Sid>
+        struct sid_neighbor_table {
+            Sid sid;
+        };
+
+        template <class IndexDimension, class NeighborDimension, int32_t MaxNumNeighbors, class Sid>
+        auto neighbor_table_neighbors(
+            sid_neighbor_table<IndexDimension, NeighborDimension, MaxNumNeighbors, Sid> const &table, size_t index) {
+            using element_type = sid::element_type<Sid>;
+
+            const auto ptr = sid_get_origin(table.sid);
+            const auto strides = sid_get_strides(table.sid);
+
+            const auto index_stride = at_key<IndexDimension>(strides);
+            const auto neighbour_stride = at_key<NeighborDimension>(strides);
+
+            gridtools::array<element_type, MaxNumNeighbors> neighbors;
+            for (int32_t elementIdx = 0; elementIdx < MaxNumNeighbors; ++elementIdx) {
```

Comment:
We should probably keep your implementation.

Reply:
I think the compiler will do its thing, but we can always check the disassembly and intervene. What I'd like to get rid of is mutating the data structure. Probably neither are worth it if we need to add template magic instead of a simple loop.